<a href="https://colab.research.google.com/github/tinawanggg/Wang_DSPN_S26/blob/master/book/exercises/03_data-as-objects-and-architectures.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 3: Data objects

These exercises are designed to get you comfortable extracting information from data objects.

We'll work with the `Credit` dataset which comes in the `ISLR` package in R. This is a simulated dataset that provides credit and demographic information on 10,000 hypothetical customers.

---
## 1. Load packages, data, model (1 point)

Install and load `ISLR` below.

In [4]:
install.packages("ISLR")
library(ISLR)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



Take a look at the first few rows of the `Credit` dataset.

In [5]:
head(Credit)

ID Income  Limit Rating Cards Age Education Gender Student Married Ethnicity
1 1   14.891 3606  283    2     34  11         Male  No      Yes     Caucasian
2 2  106.025 6645  483    3     82  15        Female Yes     Yes     Asian    
3 3  104.593 7075  514    4     71  11         Male  No      No      Asian    
4 4  148.924 9504  681    3     36  11        Female No      No      Asian    
5 5   55.882 4897  357    2     68  16         Male  No      Yes     Caucasian
6 6   80.180 8047  569    4     77  10         Male  No      No      Caucasian
  Balance
1  333   
2  903   
3  580   
4  964   
5  331   
6 1151

We can see that we have a nice **tidy** data frame here. Each column is a separate variable and each row is a different observation (in this case, simulated customers).

The code below fits a linear model to predict credit card balance from the card limit and the card owner's credit rating, age, gender, and student status. This model is saved as the `cred_lm` model object. The `summary()` function extracts important summary information from the model object so we can interpret the results.

In [7]:
cred_lm  <- lm(Balance ~ Limit + Rating + Age + Gender + Student, Credit)
summary(cred_lm)


Call:
lm(formula = Balance ~ Limit + Rating + Age + Gender + Student, 
    data = Credit)

Residuals:
    Min      1Q  Median      3Q     Max 
-682.49 -127.65    3.92  135.17  453.60 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -292.49382   48.88704  -5.983 4.92e-09 ***
Limit           0.05265    0.05392   0.976 0.329428    
Rating          1.80924    0.80462   2.249 0.025092 *  
Age            -2.12967    0.57212  -3.722 0.000226 ***
GenderFemale   -0.34809   19.63579  -0.018 0.985865    
StudentYes    397.16094   32.76189  12.123  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 195.9 on 394 degrees of freedom
Multiple R-squared:  0.8207,	Adjusted R-squared:  0.8184 
F-statistic: 360.6 on 5 and 394 DF,  p-value: < 2.2e-16


## 2. Replicating `summary` outputs (5 pts)

Let's see if we can replicate some of the values included in the `summary()` output.

Let's start with the residual standard error, aka sigma. We can see above that this is 195.9 for this model. You can directly extract sigma as follows:

In [8]:
sigma(cred_lm)

[1] 195.9206

In `lm`, sigma is calculated as

$$ \sqrt{\frac{SSE}{n-p}} $$

Where SSE is the sum of squared errors, `n` is the number of observations, and `p` is the number of parameters estimated (hint: this includes the intercept). So the denominator boils down to the degrees of freedom.

Below, use what you've learned about extracting information from model objects to calculate the SSE and extract `n` and `p`.

Hint: remember that R is really good at *vectorized operations*, meaning it easily applies the same operation individually to each element of a given vector.


---
---
### If you use Generative AI (optional)
If you use code-assisting generative AI (e.g., Copilot, GPT, Claude; more than simple google searching), follow the steps below. If you **do not** use generative AI, skip this section and go directly to the coding part.

Your workflow should be
1. Write pseudocode
2. Generate code using generative AI
3. Compare and verify the code


**Definition**: Pseudocode is a short description of an algorithm written in plain, structured steps that a human (or an AI!) can easily read and translate into real code. Pseudocode usually does not run as. It does not need to be perfect—its purpose is to help you organize your approach and to provide clear instructions to the AI.

You may write pseudocode as comments (starting with #) inside a code cell or as plain text.

#### Step 1: Write pseudocode

Write pseudocode for `SSE`, `n`, and `p` below.

In [ ]:
# Write out your pseudo code here

#### Step 2: Generate code using generative AI
Using your pseudocode, prompt a generative AI to create or fix the R functionso it follows the instructions. Then paste **only the generated function** below (not the full AI response).

In [ ]:
# Paste generated function here

#### Step 3: Compare and verify the code

Compare your pseudocode and the AI-generated code. Answer the questions below concisely (a few bullet points is enough.)

**Question 1.** What is different between your pseudocode and generated code.
(e.g., it handled NAs differently, used a different formula arrangement, etc.)

**Question 2.** What are the input parameters and the output of the function?
(Answer to show you understand what the function takes in and returns. Be concise.)

**Question 3.** Does anything has to be changed? (Yes / No)

Write your answer here
##### Question 1.
-

##### Question 2.
-
##### Question 3.
- Yes / No

If you answers yes to the question 3, revise the code (either using GenAI or on your own). Don't forget to paste your final version in the code block below.

---
---

In [10]:
# INSERT CODE HERE
# Calculate SSE
SSE <- sum(residuals(cred_lm)^2)

# Extract n
n <- length(residuals(cred_lm))

# Extract p
p <- length(coef(cred_lm))

Now, combine your work above to write a function that takes any fitted linear model and returns the residual standard error. Then test your function on the `cred_lm` model object. Compare your answer to sigma extracted directly from the model object.

In [ ]:
# INSERT CODE HERE
sigma2 <- function()

In [ ]:
# Test and compare results.

your_function(cred_lm) #Replace with your own function name
sigma(cred_lm)

---
## 2. Summary table and indexing (4 pts)

Let's say we wanted to extract the entire coefficient table provided to us by the `summary()` function above, maybe for use in a publication. You might expect this to be pulled by:

In [ ]:
cred_lm$coefficients

But as we saw in the tutorial, this pulls just the variable name and estimate, and not the standard error, t-statistic, or p-value. You could try to find where all this information is stored in the `cred_lm` object using the `str()` function.

In [ ]:
str(cred_lm)

But you actually won't find it in there! That's because the information in the coefficient table is a component of `summary()`, not a component of the model object itself. That's right, `summary()` creates it's own object that you can further pull information from.

Knowing this, pull the coefficient table from the `summary()` object.

In [ ]:
# INSERT CODE HERE


Maybe we are not interested in including the t-statistic in our final table. Pull **just** the estimate, SE, and p-value from the `summary()` object.

In [ ]:
# INSERT CODE HERE


Now, pull the table again but drop the `(Intercept)` term. (Don't save and alter your table above -- practice pulling the same table, minus the intercept term, directly from the summary.)

In [ ]:
# INSERT CODE HERE


That's all for Exercise 3! When you are finished, save the notebook as Exercise3.ipynb, push it to your class GitHub repository and send the instructors a link to your notebook via Canvas. You can send messages via Canvas by clicking "Inbox" on the left and then pressing the icon with a pencil inside a square.

**DUE:** 5pm EST, Feb 3, 2026

**IMPORTANT** Did you collaborate with anyone on this assignment? If so, list their names here.
> *N/A*